planetarium/NineChronicles
# Feature/headless tool

### Description

This feature sets up local headless runner for unity editor and run when unity editor is on play mode.


### How to test

1. Pull this branch to your local repository
2. Setup headless runner by `Tools > Headless > Setup NineChronicles.Headless repository` menu.
  - This clones another copy of NineChronicles.Headless repository.
  - This shares Lib9c of NineChronicles by symlink to apply change immediately.
3. Setup genesis block by `Tools > Headless >  Delete and re-create genesis block` menu.
  - This also deletes old blockchain store data
  - This is the very same action with `Tools > Libplanet > Delete All (Editor)` menu.
4. Setup `appsettings.local.json` by `Tools > Headless > Setup headless config` menu.
  - This action sets `appsettings.local.json` onto newly cloned headless repository.
5. Check `Use local headless` checkbox on Game scene just like following image
  - ![Screenshot from 2023-03-28 00-48-05](https://user-images.githubusercontent.com/3654082/227994291-699ad9a8-4f81-41a3-8abb-f49d3e1326b8.png)
6. Play Unity editor
  - Another process will start and run local headless.
  - Chain data will be lost because default store type is `memory`.
    - You can change this by editing `HeadlessHelper.cs`. find `--store-type memory` and change it to `--store-type rocksdb`.

### TODO
- [ ] Symlink for all OS
  - [ ] Test for OSX
  - [ ] Test for Linux
- [ ] Use mainnet snapshot as store path
- [ ] Use mainnet genesis block instead of creating new one
- [ ] Check tutorial problem by `UNITY_EDITOR` config. (`GameConfig.cs`?)
- [ ] (Optional) Create genesis block by headless function
